src/Raven.Client/Documents/Session/InMemoryDocumentSessionOperations.cs

```diff
@@ -1931,4 +1940,215 @@ public void Dispose()
+
+            _onBeforeStoreDocumentsByEntity = new Dictionary<object, DocumentInfo>(ObjectReferenceEqualityComparer<object>.Default);
+        }
+
+        public void Clear()
+        {
+            _documentsByEntity.Clear();
+            _onBeforeStoreDocumentsByEntity?.Clear();
+        }
+
+        public bool TryGetValue(object entity, out DocumentInfo documentInfo)
+        {
+            if (_documentsByEntity.TryGetValue(entity, out documentInfo))
+                return true;
+
+            return _onBeforeStoreDocumentsByEntity != null && _onBeforeStoreDocumentsByEntity.TryGetValue(entity, out documentInfo);
+        }
+
+        public IEnumerator<DocumentsByEntityEnumeratorResult> GetEnumerator()
+        {
+            foreach (var doc in _documentsByEntity)
+            {
+                yield return new DocumentsByEntityEnumeratorResult
+                {
+                    Key = doc.Key,
+                    Value = doc.Value,
+                    ExecuteOnBeforeStore = true
+                };
+            }
+
+            if (_onBeforeStoreDocumentsByEntity != null)
+            {
+                foreach (var doc in _onBeforeStoreDocumentsByEntity)
+                {
+                    yield return new DocumentsByEntityEnumeratorResult
+                    {
+                        Key = doc.Key,
+                        Value = doc.Value,
+                        ExecuteOnBeforeStore = false
+                    };
+                }
+            }
+        }
+
+        public IDisposable PrepareEntitiesPuts()
+        {
+            _prepareEntitiesPuts = true;
+
+            return new DisposableAction(() => _prepareEntitiesPuts = false);
+        }
+
+        internal class DocumentsByEntityEnumeratorResult
+        {
+            public object Key { get; set; }
+
+            public DocumentInfo Value { get; set; }
+
+            public bool ExecuteOnBeforeStore { get; set; }
+        }
+    }
+    
+    internal class DeletedEntitiesHolder
+    {
+        private readonly HashSet<object> _deletedEntities = new HashSet<object>(ObjectReferenceEqualityComparer<object>.Default);
+
+        private HashSet<object> _onBeforeDeletedEntities;
+
+        private bool _prepareEntitiesDeletes;
+
+        public int Count => _deletedEntities.Count + _onBeforeDeletedEntities?.Count ?? 0;
+
+        public void Add(object entity)
+        {
+            if (_prepareEntitiesDeletes)
+            {
+                if (_onBeforeDeletedEntities == null)
+                    _onBeforeDeletedEntities = new HashSet<object>(ObjectReferenceEqualityComparer<object>.Default);
+
+                _onBeforeDeletedEntities.Add(entity);
+                return;
+            }
+
+            _deletedEntities.Add(entity);
+        }
+
+        public void Remove(object entity)
+        {
+            _deletedEntities.Remove(entity);
+            _onBeforeDeletedEntities?.Remove(entity);
+        }
+
+        public void Evict(object entity)
```

Comment:
Why does evict goes here?

Reply:
Because we cannot Evict during `OnBeforeStore`